cxx4opencl/address_spaces.txt

```diff
@@ -40,63 +91,73 @@ Address spaces are not deduced for:
   * non-pointer/non-reference alias declarations.
   * decltype expressions.
 
-[source,c]
+[source,cpp]
 ----------
 template <typename T>
 void foo() {
-  T m; // address space of m will be known at template instantiation time.
-  T * ptr; // ptr points to generic address space object.
-  T & ref = ...; // ref references an object in generic address space.
+  T m; // address space of 'm' will be known at template instantiation time.
+  T * ptr; // 'ptr' points to generic address space object.
+  T & ref = ...; // 'ref' references an object in generic address space.
 };
 
 template <int N>
 struct S {
-  int i; // i has no address space.
-  static int ii; // ii is in global address space.
-  int * ptr; // ptr points to generic address space int.
-  int & ref = ...; // ref references int in generic address space.
+  int i; // 'i' has no address space.
+  static int ii; // 'ii' is in global address space.
+  int * ptr; // 'ptr' points to generic address space int.
+  int & ref = ...; // 'ref' references int in generic address space.
 };
 
 template <int N>
 void bar()
 {
-  S<N> s; // s is in __private address space.
+  S<N> s; // 's' is in __private address space.
 }
 ----------
 
-TODO: Example for type alias and decltype!
-
-==== References
-
-Reference types can be qualified with an address space.
-
-[source,c]
+[source,cpp]
 ----------
-__private int & ref = ...; // references int in __private address space.
+struct c1 {};
+using alias_c1 = c1; // 'alias_c1' is 'c1'.
+using alias_c1_ptr = c1 *; // 'alias_c1_ptr' is a generic address space pointer to 'c1'.
 ----------
 
-By default references will refer to generic address space objects, except
-for dependent types that are not template specializations (see
-<<addrspace-deduction, _Deduction_>>). Address space compatibility checks
-are performed when references are bound to values. The logic follows the
-rules from address space pointer conversion (`OpenCL v2.0 s6.5.5`).
+[source,cpp]
+----------
+__kernel void foo()
+{
+  __local int i;
+  decltype(i)* ii; // type of 'ii' is '__local int *__private'.
+}
+----------
 
-==== Default address space
+For the placeholder type specifier `auto` an address space of the outer type is
+deduced as if it would be any other regular type. However if auto is used in a
+reference or pointer type, the address space of a pointee is taken from the type
+of the initialization expression. The logic follows `const`/`volatile`
+qualifiers.
```

Comment:
```suggestion
qualifier logic.
```

Reply:
There is already "logic" at the beginning of this sentence. I think I will change to rules.

Reply:
That's fine with me too.

Reply:
https://github.com/KhronosGroup/OpenCL-Docs/pull/263/commits/079a393b2ba07a6d68f1292a15b93aa514cd2276